Firestore/Source/Local/FSTMemoryQueryCache.mm

```diff
@@ -113,11 +118,54 @@ - (nullable FSTQueryData *)queryDataForQuery:(FSTQuery *)query {
   return self.queries[query];
 }
 
+- (void)enumerateSequenceNumbersUsingBlock:(void (^)(FSTListenSequenceNumber sequenceNumber,
+                                                     BOOL *stop))block {
+  [self.queries
+      enumerateKeysAndObjectsUsingBlock:^(FSTQuery *key, FSTQueryData *queryData, BOOL *stop) {
+        block(queryData.sequenceNumber, stop);
+      }];
+  [self.mutatedDocumentSequenceNumbers
+      enumerateKeysAndObjectsUsingBlock:^(FSTDocumentKey *key, NSNumber *sequenceNumber,
+                                          BOOL *stop) {
+        block([sequenceNumber longLongValue], stop);
+      }];
+}
+
+- (NSUInteger)removeQueriesThroughSequenceNumber:(FSTListenSequenceNumber)sequenceNumber
+                                     liveQueries:
+                                         (NSDictionary<NSNumber *, FSTQueryData *> *)liveQueries
+                                           group:(__unused FSTWriteGroup *)group {
+  NSMutableArray<FSTQuery *> *toRemove = [NSMutableArray array];
+  [self.queries
+      enumerateKeysAndObjectsUsingBlock:^(FSTQuery *query, FSTQueryData *queryData, BOOL *stop) {
+        if (queryData.sequenceNumber <= sequenceNumber) {
+          if (liveQueries[@(queryData.targetID)] == nil) {
+            [toRemove addObject:query];
+          }
+        }
+      }];
+  [self.queries removeObjectsForKeys:toRemove];
```

Comment:
Do we also need to remove document keys from the mutated document sequence numbers?